src/devices/machine/cat702.cpp

```diff
@@ -231,3 +237,63 @@ void cat702_device::write_datain(int state)
 {
 	m_datain = state;
 }
+
+void cat702_piu_device::write_select(int state)
+{
+	if (m_select != state)
+	{
+		if (!state)
+		{
+			m_state = 0xfc;
+			m_bit = 0;
+
+			apply_sbox(initial_sbox);
+		}
+		else
+		{
+			m_dataout_handler(1);
+		}
+
+		m_select = state;
+	}
+}
+
+void cat702_piu_device::write_clock(int state)
+{
+	/*
+	Pump It Up uses a CAT702 but accesses it directly and in a way that
+	seems conflicting with how the ZN uses it through the PS1's SIO.
+
+	This is the sequence performed with clock and data lines write and read data:
+	write unkbit 0
+	write unkbit 1
+	write select 0
+	loop for all bits that need to be transferred:
+	    write clk = 0, data = x
+	    write clk = x, data = 0/1
+	    write clk = 1, data = x
+	    read data
+	write select 1
+
+	Modifying the old code to work with PIU breaks ZN games.
+	*/
```

Comment:
Which part specifically upsets ZN1?  I can see two changes here:
* Doing both input and output on the rising clock edge, while the existing class does output on the falling edge and input on the rising edge.
* Applying the initial S-box on select being asserted, rather than doing it on a falling clock edge when the bit counter is zero.

Does ZN1 depend on the S-box being reset when transferring multiple bytes without deasserting select in between, which PIU depends on the S-box *not* being reset in this situation?

Reply:
I can't tell you anything about what ZN1 games expect because I'm not working with ZN1 beyond testing to see if any changes were breaking on there.

I've written out the exact order of how the PIU works with the CAT702 device and it already breaks in the very first clock write if you work through the states using the original code's logic.

Assuming the first clock write is 0, then `apply_sbox(initial_sbox)` is called and it outputs the first result bit to the dataout callback which is not ok because no data has been written or applied yet. In this case, the real output bit that PIU expects to read can't be read until the start of the next loop but PIU is expecting to be able to read the output bit immediately at the end before looping again.

If you assume the first clock write should be 1 instead of 0 then there's a different problem. m_clock defaults to 1 so the first clock would be ignored, and even if it wasn't ignored it still hasn't written the data bit yet so the data bit won't be applied until the next loop, and it would also increase m_bit so `apply_sbox(initial_sbox)` wouldn't be applied on the first byte. So this also goes out of sync immediately here.

There's just a fundamental difference in how PIU wants to use the device (which imo is a pretty straight forward direct access usage) vs how the ZN1 is expecting to be able to use it via the PS1's SIO.

Reply:
Right, but will PIU tolerate something like:

```c++
void cat702_piu_device::write_select(int state)
{
	if (m_select != state)
	{
		if (!state)
		{
			m_state = 0xfc;
			m_bit = 0;
		}
		else
		{
			m_dataout_handler(1);
		}

		m_select = state;
	}
}

void cat702_piu_device::write_clock(int state)
{
	/*
	Pump It Up uses a CAT702 but accesses it directly and in a way that
	seems conflicting with how the ZN uses it through the PS1's SIO.

	This is the sequence performed with clock and data lines write and read data:
	write unkbit 0
	write unkbit 1
	write select 0
	loop for all bits that need to be transferred:
	    write clk = 0, data = x
	    write clk = x, data = 0/1
	    write clk = 1, data = x
	    read data
	write select 1

	Modifying the old code to work with PIU breaks ZN games.
	*/
	if (!state && m_clock && !m_select)
	{
		apply_sbox(initial_sbox);
	}

	if (state && !m_clock && !m_select)
	{
		// Compute the output and change the state
		if (!m_datain)
			apply_bit_sbox(m_bit);

		m_bit++;
		m_bit&=7;

		if (m_bit==0)
		{
			// Apply the initial sbox
			apply_sbox(initial_sbox);
		}

		m_dataout_handler(((m_state >> m_bit) & 1) != 0);
	}

	m_clock = state;
}
```

or does it depend on the S-box not being reset between consecutive byte transfers while select remains asserted?

Reply:
Doing something like that would just give completely different outputs. apply_sbox is completely mutating m_state based on the current value of m_state using the provided s-box so it's really just a full state mutation and can only happen at the one timing when m_bit is 0.

Reply:
Sorry, screwed up editing the code in the comment, and now I can’t work out what I was thinking.

One way or another, I suspect zn.cpp is depending on something wrong somewhere in the chain, even if it isn’t immediately clear what that is.